cli/src/cast.rs

```diff
@@ -175,6 +175,26 @@ async fn main() -> eyre::Result<()> {
             let tokens = utils::format_tokens(&tokens);
             tokens.for_each(|t| println!("{}", t));
         }
+        Subcommands::FourByte { selector } => {
+            let sigs = SimpleCast::fourbyte(&selector).await?;
+            sigs.iter().for_each(|sig| println!("{}", sig));
+        }
+        Subcommands::FourByteDecode { calldata } => {
+            let sigs = SimpleCast::fourbyte(&calldata).await?;
+            sigs.iter().enumerate().for_each(|(i, sig)| println!("{}) \"{}\"", i + 1, sig));
+            print!("Select a function signature by number: ");
+            io::stdout().flush()?;
+
+            let mut input = String::new();
+            io::stdin().read_line(&mut input)?;
+            let i: usize = input.trim().parse()?;
```

Comment:
Is this necessary? When would we expect multiple function signatures to match the 4 bytes?

Reply:
This happens pretty frequently actually, as shown below with some common selectors.

When [implementing](https://github.com/dapphub/dapptools/blob/465aab0ff637441ea39183b090b124b6eb862a88/src/seth/libexec/seth/seth-4byte-decode#L11-L41) this in seth we decided to prompt the user to choose a selector to decode with, but we also had a `SETH_4BYTE_ID` env var which, if specified, would choose the selector with that ID non-interactively (the IDs are from 4byte.directory's DB, see second code block). `earliest` and `latest` were valid values here so you could assume e.g. ERC20 sigs are the earliest ones added to the DB to non-interactively select them

Another idea which was never implemented in seth is to use the calldata to non-interactively choose a signature, when it's unambiguous which sig to use. For example, this is easy to figure out for `0x23b872dd` if the calldata has three params, since we know it can't be the `int128` sig in that case

```
$ seth 4byte 0xa9059cbb
func_2093253501(bytes)
transfer(bytes4[9],bytes5[6],int48[11])
many_msg_babbage(bytes1)
transfer(address,uint256)

$ seth 4byte 0x23b872dd
gasprice_bit_ether(int128)
transferFrom(address,address,uint256)

$ seth 4byte 0x01ffc9a7
pizza_mandate_apology(uint256)
supportsInterface(bytes4)
```

```
$ curl -s https://www.4byte.directory/api/v1/signatures/\?hex_signature\=0x23b872dd | jq .
{
  "count": 2,
  "next": null,
  "previous": null,
  "results": [
    {
      "id": 31781,
      "created_at": "2018-05-12T20:40:45.467194Z",
      "text_signature": "gasprice_bit_ether(int128)",
      "hex_signature": "0x23b872dd",
      "bytes_signature": "#¸rÝ"
    },
    {
      "id": 147,
      "created_at": "2016-07-09T03:58:28.927638Z",
      "text_signature": "transferFrom(address,address,uint256)",
      "hex_signature": "0x23b872dd",
      "bytes_signature": "#¸rÝ"
    }
  ]
}
```

Reply:
All great ideas. 

So:
1. We'd add the env var which lets you skip the interactive step
2. We'd check length of args provided and filter out any of the found signatures with non-matching lengths, to trim down the number of choices. If that gets us to 1 choice, then the command should be non-interactive

@ncitron up to you if you want these to happen in this PR or not, these sound valuable enough that we should add them now perhaps?

Reply:
Ah one more thing is we should default to non-interactive if only one sig is returned, which we don't currently do. So the behavior of `4byte-decode` should be something like this:

1. Query 4byte.directory
2. If we only receive one sig, decode with that and return. (And print warning if env var is specified and doesn't match the returned sig?)
3. If we receive multiple sigs, check if env var is specified
4. If env var is specified and valid, use that and return
4. If env var doesn't match any IDs, print warning and continue to next step
5. Try automatically decoding based on arg length. If unambiguous, decode and return
6. Lastly, prompt user to select a sig 

@gakonst feel free to tweak the above but that sounds like the best UX to me

Reply:
Yeah I can handle these changes. Might take me a day or two.

For the offline database, would it be better to have a --offline flag rather than an environment variable?

Reply:
Good call, I think an `--id` flag probably makes more sense than an env var and is more consistent with the rest of the foundry interface. In seth/dapptools most options were handled via env vars but I think foundry is moving away from that, though I'll let @gakonst confirm here

------------------------

One other thought—this process of decoding calldata without an ABI will be useful when tracing/debugging mainnet transactions with unverified contracts (e.g. `seth run-tx <txHash> --trace` functionality), so we might want to pull the logic out somewhere so it can be reused easily

Reply:
Agree with all, we don't need to be locked in env vars, and cli flags make sense.

We can put this ABI inference logic in the `utils/` crate, and import it where needed.